File: sdk-typescript/test/iwfcompat/execute_only_flow.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import {
  StepDef,
  doubleCodec,
  goTo,
  gracefulComplete,
  type Context,
  type Flow,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class ExecuteOnlySecondStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public getStepType(): string {
    return "ExecuteOnlySecondStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return gracefulComplete(input + 1);
  }
}

class ExecuteOnlyFirstStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(private readonly second: ExecuteOnlySecondStep) {}

  public getStepType(): string {
    return "ExecuteOnlyFirstStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return goTo(this.second, input + 1);
  }
}

export class ExecuteOnlyFlow implements Flow<number> {
  private readonly second = new ExecuteOnlySecondStep();
  private readonly first = new ExecuteOnlyFirstStep(this.second);

  public getFlowType(): string {
    return "ExecuteOnlyFlow";
  }

  public getSteps() {
    return [StepDef.startStep(this.first), StepDef.nonStartStep(this.second)];
  }
}

File: sdk-python/tests/iwfcompat/execute_only_flow.py
# Copyright (c) 2026 Super Durable, Inc.
#
# Licensed under the Super Durable Source License 1.0.
# You may not use this file except in compliance with the License.
# See the LICENSE file in the repository root.
#
# SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

from dex import Context, Flow, Step, StepDecision, StepDef, go_to, graceful_complete


class ExecuteOnlySecondStep(Step[int]):
    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return graceful_complete(input + 1)


class ExecuteOnlyFirstStep(Step[int]):
    def __init__(self, second: ExecuteOnlySecondStep) -> None:
        self.second = second

    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return go_to(self.second, input + 1)


class ExecuteOnlyFlow(Flow[int]):
    def __init__(self) -> None:
        self.second = ExecuteOnlySecondStep()
        self.first = ExecuteOnlyFirstStep(self.second)

    def get_steps(self) -> tuple[StepDef, ...]:
        return (
            StepDef.start_step(self.first),
            StepDef.non_start_step(self.second),
        )
